fix: Provide future to get the enrollmentStatus and refactor the AtEn…

sitaram-kalluri opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- Refactor submitEnrollmentRequest method to return a String type which represents an enrollment request id.
- Modify submitEnrollmentRequest method to send only one enrollment request and reject enrollment request until the pending enrollment request method is fulfilled.
- Mark the "initEnrollmentAuthScheduler" as a private method and expose "getFinalEnrollmentStatus" method which will invoke the "initEnrollmentAuthScheduler" method and returns a future which will complete with the final enrollment status.

**- How I did it**
- The "submitEnrollmentRequest" is modified to return an enrollment request id instead of "AtEnrollmentResponse". Also remove the call that invokes the "initEnrollmentAuthScheduler". 
- In "submitEnrollmentRequest" to verify if there is a pending enrollment request. If yes, throw "InvalidRequestException".
- Introduce "approve" and "deny" method to approve/deny the enrollment request and move the internal logic to the private method and invoke from the above public methods.

**- How to verify it**
- Manually testing the changes with the mobile apps.

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

murali-shris committed:
fix: APKAM Enrollment changes

sitaram-kalluri committed:
fix: Update getEncryptedKeys method to add enrollment id in backup keys

sitaram-kalluri committed:
fix: Implement isOnboarded method

murali-shris committed:
feat: add functional tests for apkam

sitaram-kalluri committed:
fix: Prefix '@' to the atsign

sitaram-kalluri committed:
feat: Introduce methods to set SPP and fetch OTP from the secondary server

srieteja committed:
feat: introduce fetchEnrollmentRequest() and underlying classes

srieteja committed:
test: add unit test

srieteja committed:
test: add functional tests

srieteja committed:
feat: add new classes to the list of exports

srieteja committed:
docs: update pubspec version and changelog

sitaram-kalluri committed:
fix: Set enrollmentId in at_client_impl.dart and fix unit tests

dependabot[bot] committed:
build(deps): bump codecov/codecov-action from 3.1.4 to 3.1.5

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3.1.4 to 3.1.5.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/eaaf4bedf32dbdc6b720b63067d99c4d77d6047d...4fe8c5f003fae66aa5ebb77cfd3e7bfbbda0b6b0)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1224 from atsign-foundation/dependabot/github_actions/codecov/codecov-action-3.1.5

build(deps): bump codecov/codecov-action from 3.1.4 to 3.1.5

srieteja committed:
Update CHANGELOG.md

srieteja committed:
Merge branch 'trunk' into enroll_list_client_2

srieteja committed:
Update CHANGELOG.md

murali-shris committed:
Merge pull request #1222 from atsign-foundation/enroll_list_client_2

feat: add new method to fetch enrollment requests

dependabot[bot] committed:
build(deps): bump dart-lang/setup-dart from 1.6.1 to 1.6.2

Bumps [dart-lang/setup-dart](https://github.com/dart-lang/setup-dart) from 1.6.1 to 1.6.2.
- [Release notes](https://github.com/dart-lang/setup-dart/releases)
- [Changelog](https://github.com/dart-lang/setup-dart/blob/main/CHANGELOG.md)
- [Commits](https://github.com/dart-lang/setup-dart/compare/ca7e6fee45ffbd82b555a7ebfc236d2c86439f5b...fedb1266e91cf51be2fdb382869461a434b920a3)

---
updated-dependencies:
- dependency-name: dart-lang/setup-dart
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump codecov/codecov-action from 3.1.5 to 3.1.6

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3.1.5 to 3.1.6.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/4fe8c5f003fae66aa5ebb77cfd3e7bfbbda0b6b0...ab904c41d6ece82784817410c45d8b8c02684457)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump step-security/harden-runner from 2.6.1 to 2.7.0

Bumps [step-security/harden-runner](https://github.com/step-security/harden-runner) from 2.6.1 to 2.7.0.
- [Release notes](https://github.com/step-security/harden-runner/releases)
- [Commits](https://github.com/step-security/harden-runner/compare/eb238b55efaa70779f274895e782ed17c84f2895...63c24ba6bd7ba022e95695ff85de572c04a18142)

---
updated-dependencies:
- dependency-name: step-security/harden-runner
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
build(deps): Rollup merge branch for #1228 dependabot/github_actions/step-security/harden-runner-2.7.0

cpswan committed:
build(deps): Rollup merge branch for #1227 dependabot/github_actions/codecov/codecov-action-3.1.6

cpswan committed:
Merge pull request #1226 from atsign-foundation/dependabot/github_actions/dart-lang/setup-dart-1.6.2

build(deps): bump dart-lang/setup-dart from 1.6.1 to 1.6.2

sitaram-kalluri committed:
fix: Uptake at_commons 4.0.0 changes

dependabot[bot] committed:
build(deps): bump peter-evans/create-pull-request from 5.0.2 to 6.0.0

Bumps [peter-evans/create-pull-request](https://github.com/peter-evans/create-pull-request) from 5.0.2 to 6.0.0.
- [Release notes](https://github.com/peter-evans/create-pull-request/releases)
- [Commits](https://github.com/peter-evans/create-pull-request/compare/153407881ec5c347639a548ade7d8ad1d6740e38...b1ddad2c994a25fbc81a28b3ec0e368bb2021c50)

---
updated-dependencies:
- dependency-name: peter-evans/create-pull-request
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1229 from atsign-foundation/dependabot/github_actions/peter-evans/create-pull-request-6.0.0

build(deps): bump peter-evans/create-pull-request from 5.0.2 to 6.0.0

dependabot[bot] committed:
build(deps): bump codecov/codecov-action from 3.1.6 to 4.0.0

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3.1.6 to 4.0.0.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/ab904c41d6ece82784817410c45d8b8c02684457...f30e4959ba63075080d4f7f90cacc18d9f3fafd7)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

gkc committed:
Merge pull request #1230 from atsign-foundation/dependabot/github_actions/codecov/codecov-action-4.0.0

build(deps): bump codecov/codecov-action from 3.1.6 to 4.0.0

dependabot[bot] committed:
build(deps): bump codecov/codecov-action from 4.0.0 to 4.0.1

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 4.0.0 to 4.0.1.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/f30e4959ba63075080d4f7f90cacc18d9f3fafd7...e0b68c6749509c5f83f984dd99a76a1c1a231044)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1231 from atsign-foundation/dependabot/github_actions/codecov/codecov-action-4.0.1

build(deps): bump codecov/codecov-action from 4.0.0 to 4.0.1

sitaram-kalluri committed:
fix: Provide future to get the enrollmentStatus and refactor the AtEnrollmentService

sitaram-kalluri committed:
fix: Provide future to get the enrollmentStatus and refactor the AtEnrollmentService

sitaram-kalluri committed:
fix: Include exceptions in dart docs and update at_lookup package version

sitaram-kalluri committed:
Merge branch 'trunk' into 1219-introduce-methods-to-set-spp-and-fetch-otp-in-at_client-sdk

sitaram-kalluri committed:
fix: Fix enrollment_test.dart

sitaram-kalluri committed:
fix: Set the enrollmentId in AtClient._() method

dependabot[bot] committed:
build(deps): bump actions/upload-artifact from 4.3.0 to 4.3.1

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 4.3.0 to 4.3.1.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/26f96dfa697d77e81fd5907df203aa23a56210a8...5d5d22a31266ced268874388b861e4b58bb5c2f3)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1233 from atsign-foundation/dependabot/github_actions/actions/upload-artifact-4.3.1

build(deps): bump actions/upload-artifact from 4.3.0 to 4.3.1

purnimavenkatasubbu committed:
updated dependencies

purnimavenkatasubbu committed:
added at_client version in changelog

sitaram-kalluri committed:
Merge branch 'trunk' into 1219-introduce-methods-to-set-spp-and-fetch-otp-in-at_client-sdk

sitaram-kalluri committed:
Merge pull request #1223 from atsign-foundation/1219-introduce-methods-to-set-spp-and-fetch-otp-in-at_client-sdk

feat: Introduce methods to set SPP and fetch OTP from the secondary server

gkc committed:
Merge branch 'trunk' into issue_1232

cpswan committed:
ci: Add unit and functional tests using Dart beta

cpswan committed:
fix: Workflow name - add _dart_beta suffix

cpswan committed:
ci: Use matrix for Dart channel

gkc committed:
Merge pull request #1236 from atsign-foundation/cpswan-beta-tests

sitaram-kalluri committed:
fix: Fix intermittent failing compaction job test in functional test

gkc commented on packages/at_client_mobile/lib/src/enrollment/at_enrollment_service.dart:
```diff
@@ -1,41 +1,31 @@
-  void initEnrollmentAuthScheduler();
+  /// [EnrollmentStatus.denied] indicates that the enrollment ID is not eligible for
+  /// APKAM authentication.
+  Future<EnrollmentStatus> getFinalEnrollmentStatus();
```
Can you rename to `get finalEnrollmentStatus` please

sitaram-kalluri replied:
Renamed the method as `get finalEnrollmentStatus`

gkc commented on packages/at_client_mobile/lib/src/enrollment/at_enrollment_service_impl.dart:
```diff
@@ -43,15 +43,27 @@ class AtEnrollmentServiceImpl implements AtEnrollmentService {
 
   @override
-  Future<AtEnrollmentResponse> submitEnrollmentRequest(
+  Future<String> submitEnrollmentRequest(
```
Consider renaming to `submit` or `request`

sitaram-kalluri replied:
Renamed the method to `submit`

gkc commented on packages/at_client_mobile/lib/src/enrollment/at_enrollment_service_impl.dart:
```diff
@@ -287,8 +340,16 @@ class AtEnrollmentServiceImpl implements AtEnrollmentService {
+  Future<AtEnrollmentResponse> deny(AtEnrollmentRequest atEnrollmentRequest) {
+    return _manageEnrollmentApproval(atEnrollmentRequest);
+  }
+
```
This seems not quite right - approve and deny have exactly the same implementation?

sitaram-kalluri replied:
Since, the logic to approve/deny enrollment request is common between apps and the cli, the logic is placed in at_auth package.
```
In at_enrollment_impl.dart->manageEnrollmentApproval(AtEnrollmentRequest atEnrollmentRequest, AtLookUp atLookUp)
```
Link to at_enrollment_base.dart : [at_enrollment_base.dart](https://github.com/atsign-foundation/at_libraries/blob/trunk/packages/at_auth/lib/src/enroll/at_enrollment_base.dart)

In at_client_mobile.dart, for approving or denying an enrollment request, the request is dispatched to the `manageEnrollmentApproval` function within the at_auth package. Within at_client_mobile, the `_manageEnrollmentApproval` function includes a call to manageEnrollmentApproval in the at_auth package.

Since, at_auth package is used internally, can we do a breaking change to rename the methods `manageEnrollmentApproval` to approve and deny (similar to what we did in at_client_mobile)?

gkc replied:
> Since, at_auth package is used internally, can we do a breaking change to rename the methods `manageEnrollmentApproval` to approve and deny (similar to what we did in at_client_mobile)?

Don't need a breaking change; deprecate the existing method and add `approve` and `deny` methods

sitaram-kalluri replied:
Sure Gary.

dependabot[bot] committed:
build(deps): bump actions/dependency-review-action from 4.0.0 to 4.1.0

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 4.0.0 to 4.1.0.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/4901385134134e04cec5fbe5ddfe3b2c5bd5d976...80f10bf419f34980065523f5efca7ebed17576aa)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

gkc committed:
Merge pull request #1238 from atsign-foundation/dependabot/github_actions/actions/dependency-review-action-4.1.0

github-actions[bot] committed:
chore: Update cert in testdata

cpswan committed:
Merge pull request #1239 from atsign-foundation/bot-new-cert

chore: Update cert in testdata

gkc committed:
Merge branch 'trunk' into issue_1232

dependabot[bot] committed:
build(deps): bump actions/dependency-review-action from 4.1.0 to 4.1.2

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 4.1.0 to 4.1.2.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/80f10bf419f34980065523f5efca7ebed17576aa...be8bc500ee15e96754d2a6f2d34be14e945a46f3)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
chore: Dart format remvoes white space after '''

cpswan committed:
ci: Don't fail fast on unit test matrix

cpswan committed:
Merge pull request #1240 from atsign-foundation/dependabot/github_actions/actions/dependency-review-action-4.1.2

build(deps): bump actions/dependency-review-action from 4.1.0 to 4.1.2

cpswan committed:
ci: Don't fail fast on functional test matrix

gkc committed:
Merge pull request #1242 from atsign-foundation/cpswan-dont-fail-fast-functional-tests

ci: Don't fail fast on functional test matrix

sachins-geekyants committed:
feat: flutter_secure_storage migrated ot biometric_storage

sachins-geekyants committed:
feat: flutter_secure_storage migrated ot biometric_storage

dependabot[bot] committed:
build(deps): bump actions/dependency-review-action from 4.1.2 to 4.1.3

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 4.1.2 to 4.1.3.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/be8bc500ee15e96754d2a6f2d34be14e945a46f3...9129d7d40b8c12c1ed0f60400d00c92d437adcce)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1243 from atsign-foundation/dependabot/github_actions/actions/dependency-review-action-4.1.3

build(deps): bump actions/dependency-review-action from 4.1.2 to 4.1.3

gkc committed:
Merge branch 'trunk' into refactor_compaction_job

sitaram-kalluri committed:
fix: Change commitlog time interval from int to duration

sitaram-kalluri committed:
fix: Dart analyze issues

gkc committed:
Merge pull request #1244 from atsign-foundation/refactor_compaction_job

fix: Fix intermittent failing compaction job test in functional test

purnimavenkatasubbu committed:
Merge branch 'trunk' into issue_1232

purnimavenkatasubbu committed:
changed to correct version of at_lookup in changelog

sachins-geekyants committed:
feat: atsign added in enrollment key

sachins-geekyants committed:
feat: atsign added in enrollment key

gkc committed:
Merge pull request #1234 from atsign-foundation/issue_1232

feat: at_client_mobile: uptake at_chops major version

dependabot[bot] committed:
build(deps): bump codecov/codecov-action from 4.0.1 to 4.0.2

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 4.0.1 to 4.0.2.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/e0b68c6749509c5f83f984dd99a76a1c1a231044...0cfda1dd0a4ad9efc75517f399d859cd1ea4ced1)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

cpswan committed:
Merge pull request #1247 from atsign-foundation/dependabot/github_actions/codecov/codecov-action-4.0.2

build(deps): bump codecov/codecov-action from 4.0.1 to 4.0.2

knowsach committed:
feat: get saved enrollment data from key chain

knowsach committed:
feat: get saved enrollment data from key chain

sitaram-kalluri committed:
Merge pull request #1248 from atsign-foundation/apkam_onboarding_biometric_storage

Apkam onboarding biometric storage

sitaram-kalluri committed:
fix: Fix merge conflicts

sitaram-kalluri committed:
Merge remote-tracking branch 'origin/apkam_onboarding_changes_with_enrollment_status' into apkam_onboarding_changes_with_enrollment_status

# Conflicts:
#	packages/at_client_mobile/lib/src/auth/at_auth_service_impl.dart
#	packages/at_client_mobile/pubspec.yaml
#	tests/at_functional_test/lib/src/at_keys_initializer.dart
#	tests/at_functional_test/test/enrollment_test.dart

sitaram-kalluri commented:
The current version of PR has a key-chain which do not have support for IOS. Requested @sachins-geekyants  to replace the keychain with the one which has IOS support. The above changes reflect the same. Closing the PR for now. Will clean up the code and raise a new PR which will be inline with the enrollement refactoring that we doing in at_auth package.
